src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSession.java

```diff
@@ -381,6 +385,26 @@ public void run() {
         TimeUnit.MILLISECONDS);
   }
 
+  // return value:
+  //   true  - pend succeed
+  //   false - pend failed
+  public boolean tryPendRequest(RequestEntry entry) {
+    // double check. the first one doesn't lock the lock.
+    // Because negotiationSucceed only transfered from false to true.
+    // So if it is true now, it will not change in the later.
+    // But if it is false now, maybe it will change soon. So we should use lock to protect it.
+    if (!this.negotiationSucceed) {
+      synchronized (negotiationPendingSend) {
+        if (!this.negotiationSucceed) {
+          negotiationPendingSend.offer(entry);
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
```

Comment:
1. Your design here binds the responsibility of negotiation to ReplicaSession again. Why couldn't ReplicaSessionInterceptor hold the pending queue? 

2. negotiationPendingSend is declared final, but you offer items here.

Reply:
It's hard to hold pending queue by `ReplicaSessionInterceptor`. 
1. `ReplicaSessionInterceptor` is a singleton, which will serves all of the ReplicaSession. So if we do this like you said, we must use a coarse-grained lock, which will produce low effiency
2. If we use `ReplicaSessionInterceptor` to do this, we must implement a lot of logic in it. For example timeout handling, which will mantain a timeout pendingResponse queue and a PendingSend, and so on. It's too complicated, and there is a lot of overlap code with ReplicaSession. And we should modify ReplicaSession to support it.